File: spl-net/src/main/java/bgu/spl171/net/impl/TFTP/CommandEncoders/BroadcastCommandEncoder.java
package bgu.spl171.net.impl.TFTP.CommandEncoders;

import bgu.spl171.net.impl.TFTP.Commands.Broadcast;
import bgu.spl171.net.impl.TFTP.Commands.Command;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.ShortEncoderDecoder;

public class BroadcastCommandEncoder implements CommandEncoder {
    ShortEncoderDecoder shortEncoderDecoder = new ShortEncoderDecoder();
    @Override
    public byte[] encode(Command cmd) {
        Broadcast broadcast = (Broadcast) cmd;
        byte [] opCode = shortEncoderDecoder.encode((short)9);
        byte [] isAdded = broadcast.isAdded() ? new byte[]{(byte) 1} : new byte[]{(byte) 0};
        byte [] fileName = broadcast.getFileName().getBytes();

        return EncoderUtils.mergeByteArray(opCode,isAdded,fileName,new byte[]{0});

    }
}

File: java to c++ encoders/BroadcastCommandEncoder.cpp
#include "BroadcastCommandEncoder.h"
#include "EncoderUtils.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace CommandEncoders
					{
						using bgu::spl171::net::impl::TFTP::Commands::Broadcast;
						using bgu::spl171::net::impl::TFTP::Commands::Command;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::ShortEncoderDecoder;

						std::vector<char> BroadcastCommandEncoder::encode(Command *cmd)
						{
							Broadcast *broadcast = static_cast<Broadcast*>(cmd);
							std::vector<char> opCode = shortEncoderDecoder->encode(static_cast<short>(9));
							std::vector<char> isAdded = broadcast->isAdded() ? std::vector<char>{static_cast<char>(1)} : std::vector<char>{static_cast<char>(0)};
							std::vector<char> fileName = broadcast->getFileName().getBytes();

							return EncoderUtils::mergeByteArray(
							{
								opCode, isAdded, fileName, std::vector<char>{0}
							});

						}
					}
				}
			}
		}
	}
}
